### PR TITLE
Extract directly to the toolcache

### PR DESCRIPTION
**Pull request by henrymercer**
Adding the CodeQL Bundle to the toolcache seems to take a surprisingly long amount of time.  We haven't collected large scale telemetry yet, but here's some anecdotal numbers from looking at the logs on PR checks:

- Linux: ~10–15 secs
- macOS: ~15–20 secs
- Windows: ~2 minutes (!)

Therefore in this PR, we add a feature flag that extracts the CodeQL bundle directly to the toolcache, eliminating the copy to toolcache step completely.

I've tested the change by adding a PR check, temporarily modifying this to running 10 runs on each platform, and running this a bunch of times.  Next steps are dogfooding this change internally using the feature flag.

### Merge / deployment checklist

- [ ] Confirm this change is backwards compatible with existing workflows.
- [ ] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [ ] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

**Review by dbartol — APPROVED** [review]
LGTM. Looks like a relatively simple change, plus some plumbing to pass the destination directory through to all the places the write the files.

On a GitHub-hosted runner, it looks like we'll still add the bundle to the toolcache, even though the cache will disappear after the job. Is this just because we have to put it _somewhere_, and we might as well put it in the toolcache for consistency with the less-common self-hosted runner scenario?

**Comment by henrymercer** [comment]
> On a GitHub-hosted runner, it looks like we'll still add the bundle to the toolcache, even though the cache will disappear after the job. Is this just because we have to put it _somewhere_, and we might as well put it in the toolcache for consistency with the less-common self-hosted runner scenario?

Yes, pretty much.  It's also consistent with the situation where the GitHub-hosted runner images have caught up to the latest version of CodeQL and the tools we want are already installed.  Plus if anyone's still using a hardcoded toolcache path to find CodeQL, that'll still work.

**Review by NlightNFotis — APPROVED** [review]
Great!

Thank you for this change!

File: src/tools-download.ts [hunk 1]
```diff
@@ -213,6 +207,16 @@ async function downloadAndExtractZstdWithStreaming(
   await tar.extractTarZst(response, dest, tarVersion, logger);
 }
 
+/** Gets the path to the toolcache directory for the specified version of the CodeQL tools. */
+export function getToolcacheDirectory(version: string): string {
+  return path.join(
+    getRequiredEnvParam("RUNNER_TOOL_CACHE"),
```
[the remark this answers is not in the document]
Comment (henrymercer) [reply src/tools-download.ts:220]: Yeah, agreed we want to fail fast if something goes wrong.  I think that should be the case: `getRequiredEnvParam` will fail if the environment variable is unset, and the computation around `os.arch()` matches that in the toolcache API (https://github.com/actions/toolkit/blob/9cc30cb0d3d4bb7b4e5a59eaf4bfe029889bdbfb/packages/tool-cache/src/tool-cache.ts#L679).

File: src/tools-download.ts [hunk 2]
```diff
@@ -217,6 +218,15 @@ export function getToolcacheDirectory(version: string): string {
   );
 }
 
+export function writeToolcacheMarkerFile(
+  extractedPath: string,
+  logger: Logger,
+): void {
+  const markerFilePath = `${extractedPath}.complete`;
```
[the remark this answers is not in the document]
Comment (henrymercer) [reply src/tools-download.ts:231]: This is needed by the toolcache — it guards against the case where a process started caching a tool, but then got interrupted, leaving a corrupt version of the tool in the toolcache.

**Pull request merged**